20220831-add-type-restrictions-to-json-syntax.md

```diff
@@ -0,0 +1,489 @@
+        "member": { "assignable_relations": [{ "type": "user" }, { "type": "group", "relation": "member" }, { "type": "user:*" }] }
+      }
+    }
+  },
+] }
+```
+
+On write, we need to validate that:
+
+1. If the user is an object:
+   1. The type of the user should be in the list of assignable relations on the relation of the type of the object
+   - write(user=user:1, member, group:1); valid, the user type is in the list of assignable relations for the member relation of the group type
+   - write(user=group:2, parent, group:1); valid, the group type is in the list of assignable relations for the parent relation of the group type 
+   - write(user=group:2, member, group:1); invalid, the group type is not in the list of assignable relations for the member relation of the group type 
+   - write(user=user:1, parent, group:1); invalid, the user type is not in the list of assignable relations for the parent relation of the group type 
+
+2. If the user is a userset:
+   - write(user=group:2#member, member, group:1); valid, the (type=group, relation=member) is in the list of assignable relations for the member relation of the group type
+   - write(user=group:2#member, parent, group:1); invalid, the (type=group, relation=member) is not in the list of assignable relations for the parent relation of the group type
+   - write(user=group:2#parent, member, group:1); invalid, the (type=group, relation=parent) is not in the list of assignable relations for the member relation of the group type
+   - write(user=group:2#parent, parent, group:1); invalid, the (type=group, relation=parent) is not in the list of assignable relations for the parent relation of the group type
+
+3. If the user is `*`:
+   - `write(user=*, parent, group:1)`; invalid, the `parent` relation on the `group` type does not allow for any `*` relation
+   - `write(user=*, member, group:1)`; valid, the `member` relation on the `group` type allows for the `*` relation (and it will be considered all objects of type `user`)
+
+#### Updating ListObjects to Respect Type Restrictions
+
+ListObjects will be the first of the Relationship Query endpoints to take advantage of this new functionality. Our hope is that we can start building a more performant ListObjects endpoint using the new type restrictions.
+
+#### Updating Expand to Respect Type Restrictions
+
+Expand will need to be updated to ignore tuples that do not match the assignable tuples in the authorization model.
+
+#### Updating Check to Respect Type Restrictions
+
+THis will be the final phase and should be undertaken only once we are completely confident of how ListObjects and Expand are performing with the new functionality. Check is the core of OpenFGA, and we should be deligint in making sure it does not break and that any changes are properly communicated.
+
+When the time comes, check will be updated to ignore tuples that do not match the assignable tuples in the authorization model.
```

Comment:
So you mean that it will throw an error if the user passed in the request is not of the right type? Because if so, we're not "ignoring tuples" 🤔 we're simply rejecting more requests

Reply:
Check will not throw an error - when the time comes, it will ignore whatever is in the database that is not of a matching type to the provided authorization model id.

So if there is a tuple that says (`user=group:2#member, relation=reader, object=repo:openfga`) and another that says (`user=user:anne, relation=member, object=group:2`)

But the assignable relations on `repo`->`admin` do not include `{ "type": "group", "relation": "member" }` then the first relationship tuple will be ignored and a check on (`user=user:anne, relation=reader, object=repo:openfga`) will return false